payroll-processor-client/src/app/employee/employee-list/state/employee-list.service.ts

```diff
@@ -0,0 +1,40 @@
+import { HttpClient } from '@angular/common/http';
+import { Injectable } from '@angular/core';
+import { of } from 'rxjs';
+import { catchError } from 'rxjs/operators';
+
+import { Employee } from './employee-list.model';
+import { EmployeeListStore } from './employee-list.store';
+
+import { EnvService } from 'src/app/shared/env.service';
+
+@Injectable({ providedIn: 'root' })
+export class EmployeeListService {
+  private readonly apiUrl: string;
+
+  constructor(
+    private http: HttpClient,
+    envService: EnvService,
+    private store: EmployeeListStore,
+  ) {
+    this.apiUrl = envService.apiRootUrl;
+  }
+
+  getEmployees() {
+    this.store.setLoading(true);
+    return this.http
+      .get<Employee[]>(`${this.apiUrl}/Employees`)
+      .pipe(
+        catchError((err) => {
+          this.store.setError({
+            message: 'Could not load employees',
+          });
+          return of([]);
+        }),
+      )
+      .subscribe({
+        next: (employees) => this.store.set(employees),
+        complete: () => this.store.setLoading(false),
```

Comment:
The reason for putting the `store.setLoading(false)` here is that is covers both scenarios we could encounter - Success and Failure.

If we didn't put it here, we'd have to either put it in a `tap()` _after_ the `catchError()` or put it in both `catchError()` and a `tap()` before `catchError()`.

By putting our 'clean up' in `complete()`, we know we will always execute that code no matter what happens to the stream of data.

(This only applies to finite/one-time observables, like HTTP requests - for infinite observables, like `query.selectAll()`, we'll do things differently.)

Reply:
Again, I didn't know about `next` and `complete` but this makes since with things like the http request pipeline of a web api being able to call `next.Invoke` but not necessarily handling all branches that may execute. Good tip!

Reply:
The thing being passed to `.subscribe()` is an [Observer](http://reactivex.io/rxjs/class/es6/MiscJSDoc.js~ObserverDoc.html) and the thing being subscribed to is called an Observable.

Observers are objects with callback functions - they get to "observe" the events coming out of Observables (streams of events) with these callbacks.

Each callback is passed a certain kind of event from the stream and each of the callbacks is optional (we could `.subscribe({ error: err => ... })` and 'Observe' the errors only).

`next` handles each emitted event that isn't an error
`error` handles error events
`complete` handles the closing of the stream of Observables (meaning, no more events are going to be emitted)